### PR TITLE
process: refine exit type

### PR DESCRIPTION
**Pull request by marechal-p**
#### What it does

IProcessExitEvent can have both fields left undefined, or have both
fields defined. This commit refines the typings so that only one of the
fields should be set.

Fixes https://github.com/eclipse-theia/theia/issues/6741

#### How to test

`yarn run test`

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marechal-p** [comment]
This could break clients in regards to typings, because the new type is a bit more constrained and clients may not have accounted for that.

File: packages/process/src/node/terminal-process.spec.ts [hunk 1]
```diff
@@ -117,16 +117,16 @@ if (process.platform !== 'win32' || process.env.THEIA_PROCESS_TEST_OVERRIDE) {
...
-        }
+        };
 
         // tslint:disable-next-line:no-any
         async function checkOutput(proc: TerminalProcess, pattern?: RegExp): Promise<ProcessExit> {
             return new Promise<ProcessExit>((resolve, reject) => {
                 let output = '';
                 proc.outputStream.on('data', chunk => output += chunk);
-                proc.onExit(async exit => {
+                proc.onClose(async exit => {
```
Comment (kittaakos) [inline packages/process/src/node/terminal-process.spec.ts]: @marechal-p, why was this required? Has the API change revealed a bug or why did you have to change `onExit` to `onClose`. Thanks!

Other than this, the API changes look good to me. if you feel like it, perhaps you can update the changelog.

Comment (marechal-p) [reply packages/process/src/node/terminal-process.spec.ts]: Hm, more like my eye caught this while I was editing the type above. Essentially here we try to handle the output of a process, but registering to `onExit` is bogus since it could fire before the streams are empty.

Comment (kittaakos) [reply packages/process/src/node/terminal-process.spec.ts]: OK, I do not mind if you add this change to the PR too, in fact, it's cool, but please split up into two commits and we can merge it. Thanks!

**Comment by marechal-p** [comment]
Sorry for the wait. Done.

**Comment by marechal-p** [comment]
Partially fixed CI. Windows keeps failing with some Git error but I don't think it is related.

**Comment by marechal-p** [comment]
Please re-open if relevant sometime.
